### PR TITLE
Add the implementation for DeepLab v3 based on TFM implementation (as an experiment)

### PR DESCRIPTION
**Pull request by qlzh727**
# What does this PR do?
Not intended to merge for now.
Add the implementation for DeepLab v3 based on TFM implementation (as an experiment)

The current class is a wrapper around the TFM segmentation model.

1. The current class only contains the model building logic.
2. Most of the tunable params are currently backed in the __init__, and not exposed to user yet.
3. Verify the model building logic in the test. (will need tfm pip deps to work)

Next step:
1. Configure the input parsing logic for a dataset, eg pascol vol or cityscapes.
2. Mimic the training loop logic as model.train_step/eval_step
3. Train model e2e.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

### CODE REVIEW

**Review by martin-gorner — COMMENTED** [review]
I like the delegate approach. I have been able to call .compile and .fit on it although there are some errors. Training complains that some variables do not have gradients. Maybe something wrong in my code, or maybe something is not connected in the delegate implementation.

I have created a test notebook with a small and (hopefully) suitable dataset:
https://source.cloud.google.com/cloudml-demo-martin/martin-kerascv-test/+/master:DeeplabV3_test.ipynb

**Comment by martin-gorner** [comment]
Note: the goal here is to wrap the Model Graden model but not the Model Garden data pipeline code. Users can use Keras preprocessing layers to format a data feed appropriately and we will provide example code in the accompanying templates.

**Comment by qlzh727** [comment]
> I like the delegate approach. I have been able to call .compile and .fit on it although there are some errors. Training complains that some variables do not have gradients. Maybe something wrong in my code, or maybe something is not connected in the delegate implementation.
> 
> I have created a test notebook with a small and (hopefully) suitable dataset: https://source.cloud.google.com/cloudml-demo-martin/martin-kerascv-test/+/master:DeeplabV3_test.ipynb

Thanks Martin, I am able to reproduce the issue. Somehow the same error happens even I create the model from the tfm code directly (eg via their config/task/factory), need to dig deeper and see why that's the case.

**Comment by qlzh727** [comment]
The stack trace I got from the code lab is like below, and i think it is probably a GPU runtime error, since the layer it complains is actually the first layer in the model.

```
WARNING:tensorflow:Gradients do not exist for variables ['conv2d_60/bias:0', 'conv2d_61/bias:0', 'conv2d_59/bias:0', 'conv2d_62/bias:0', 'batch_normalization_2/gamma:0', 'batch_normalization_2/beta:0', 'batch_normalization_3/gamma:0', 'batch_normalization_3/beta:0', 'batch_normalization_4/gamma:0', 'batch_normalization_4/beta:0', 'batch_normalization_5/gamma:0', 'batch_normalization_5/beta:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?
WARNING:tensorflow:Gradients do not exist for variables ['conv2d_60/bias:0', 'conv2d_61/bias:0', 'conv2d_59/bias:0', 'conv2d_62/bias:0', 'batch_normalization_2/gamma:0', 'batch_normalization_2/beta:0', 'batch_normalization_3/gamma:0', 'batch_normalization_3/beta:0', 'batch_normalization_4/gamma:0', 'batch_normalization_4/beta:0', 'batch_normalization_5/gamma:0', 'batch_normalization_5/beta:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?
WARNING:tensorflow:Gradients do not exist for variables ['conv2d_60/bias:0', 'conv2d_61/bias:0', 'conv2d_59/bias:0', 'conv2d_62/bias:0', 'batch_normalization_2/gamma:0', 'batch_normalization_2/beta:0', 'batch_normalization_3/gamma:0', 'batch_normalization_3/beta:0', 'batch_normalization_4/gamma:0', 'batch_normalization_4/beta:0', 'batch_normalization_5/gamma:0', 'batch_normalization_5/beta:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?
WARNING:tensorflow:Gradients do not exist for variables ['conv2d_60/bias:0', 'conv2d_61/bias:0', 'conv2d_59/bias:0', 'conv2d_62/bias:0', 'batch_normalization_2/gamma:0', 'batch_normalization_2/beta:0', 'batch_normalization_3/gamma:0', 'batch_normalization_3/beta:0', 'batch_normalization_4/gamma:0', 'batch_normalization_4/beta:0', 'batch_normalization_5/gamma:0', 'batch_normalization_5/beta:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?
---------------------------------------------------------------------------
UnimplementedError                        Traceback (most recent call last)
[<ipython-input-38-06ac18af13cf>](https://localhost:8080/#) in <module>
      8     metrics=["mse"]
      9 )
---> 10 model.fit(training_dataset, epochs=1, steps_per_epoch=NB_IMAGES//BATCH_SIZE)

8 frames
[/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/execute.py](https://localhost:8080/#) in quick_execute(op_name, num_outputs, inputs, attrs, ctx, name)
     51     ctx.ensure_initialized()
     52     tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,
---> 53                                         inputs, attrs, num_outputs)
     54   except core._NotOkStatusException as e:
     55     if name is not None:

UnimplementedError: Graph execution error:

Detected at node 'deep_lab_v3_3/segmentation_model_6/res_net_6/conv2d_54/Conv2D' defined at (most recent call last):
    File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
      "__main__", mod_spec)
    File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
      exec(code, run_globals)
    File "/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py", line 16, in <module>
      app.launch_new_instance()
    File "/usr/local/lib/python3.7/dist-packages/traitlets/config/application.py", line 846, in launch_instance
      app.start()
    File "/usr/local/lib/python3.7/dist-packages/ipykernel/kernelapp.py", line 612, in start
      self.io_loop.start()
    File "/usr/local/lib/python3.7/dist-packages/tornado/platform/asyncio.py", line 132, in start
      self.asyncio_loop.run_forever()
    File "/usr/lib/python3.7/asyncio/base_events.py", line 541, in run_forever
      self._run_once()
    File "/usr/lib/python3.7/asyncio/base_events.py", line 1786, in _run_once
      handle._run()
    File "/usr/lib/python3.7/asyncio/events.py", line 88, in _run
      self._context.run(self._callback, *self._args)
    File "/usr/local/lib/python3.7/dist-packages/tornado/ioloop.py", line 758, in _run_callback
      ret = callback()
    File "/usr/local/lib/python3.7/dist-packages/tornado/stack_context.py", line 300, in null_wrapper
      return fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/tornado/gen.py", line 1233, in inner
      self.run()
    File "/usr/local/lib/python3.7/dist-packages/tornado/gen.py", line 1147, in run
      yielded = self.gen.send(value)
    File "/usr/local/lib/python3.7/dist-packages/ipykernel/kernelbase.py", line 365, in process_one
      yield gen.maybe_future(dispatch(*args))
    File "/usr/local/lib/python3.7/dist-packages/tornado/gen.py", line 326, in wrapper
      yielded = next(result)
    File "/usr/local/lib/python3.7/dist-packages/ipykernel/kernelbase.py", line 268, in dispatch_shell
      yield gen.maybe_future(handler(stream, idents, msg))
    File "/usr/local/lib/python3.7/dist-packages/tornado/gen.py", line 326, in wrapper
      yielded = next(result)
    File "/usr/local/lib/python3.7/dist-packages/ipykernel/kernelbase.py", line 545, in execute_request
      user_expressions, allow_stdin,
    File "/usr/local/lib/python3.7/dist-packages/tornado/gen.py", line 326, in wrapper
      yielded = next(result)
    File "/usr/local/lib/python3.7/dist-packages/ipykernel/ipkernel.py", line 306, in do_execute
      res = shell.run_cell(code, store_history=store_history, silent=silent)
    File "/usr/local/lib/python3.7/dist-packages/ipykernel/zmqshell.py", line 536, in run_cell
      return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py", line 2855, in run_cell
      raw_cell, store_history, silent, shell_futures)
    File "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py", line 2881, in _run_cell
      return runner(coro)
    File "/usr/local/lib/python3.7/dist-packages/IPython/core/async_helpers.py", line 68, in _pseudo_sync_runner
      coro.send(None)
    File "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py", line 3058, in run_cell_async
      interactivity=interactivity, compiler=compiler, result=result)
    File "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py", line 3249, in run_ast_nodes
      if (await self.run_code(code, result,  async_=asy)):
    File "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py", line 3326, in run_code
      exec(code_obj, self.user_global_ns, self.user_ns)
    File "<ipython-input-38-06ac18af13cf>", line 10, in <module>
      model.fit(training_dataset, epochs=1, steps_per_epoch=NB_IMAGES//BATCH_SIZE)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py", line 61, in error_handler
      return fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/training.py", line 1637, in fit
      tmp_logs = self.train_function(iterator)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/training.py", line 1233, in train_function
      return step_function(self, iterator)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/training.py", line 1217, in step_function
      outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/training.py", line 1206, in run_step
      outputs = model.train_step(data)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/training.py", line 1010, in train_step
      y_pred = self(x, training=True)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py", line 61, in error_handler
      return fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/training.py", line 560, in __call__
      return super().__call__(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py", line 61, in error_handler
      return fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py", line 1102, in __call__
      outputs = call_fn(inputs, *args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py", line 96, in error_handler
      return fn(*args, **kwargs)
    File "<ipython-input-3-d6ed53044862>", line 159, in call
      return self._tfm_model(inputs, training=training)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py", line 61, in error_handler
      return fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/training.py", line 560, in __call__
      return super().__call__(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py", line 61, in error_handler
      return fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py", line 1102, in __call__
      outputs = call_fn(inputs, *args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py", line 96, in error_handler
      return fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/official/vision/modeling/segmentation_model.py", line 64, in call
      backbone_features = self.backbone(inputs)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py", line 61, in error_handler
      return fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/training.py", line 560, in __call__
      return super().__call__(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py", line 61, in error_handler
      return fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py", line 1102, in __call__
      outputs = call_fn(inputs, *args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py", line 96, in error_handler
      return fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/functional.py", line 510, in call
      return self._run_internal_graph(inputs, training=training, mask=mask)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/functional.py", line 667, in _run_internal_graph
      outputs = node.layer(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py", line 61, in error_handler
      return fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py", line 1102, in __call__
      outputs = call_fn(inputs, *args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py", line 96, in error_handler
      return fn(*args, **kwargs)
    File "/usr/local/lib/python3.7/dist-packages/keras/layers/convolutional/base_conv.py", line 283, in call
      outputs = self.convolution_op(inputs, self.kernel)
    File "/usr/local/lib/python3.7/dist-packages/keras/layers/convolutional/base_conv.py", line 262, in convolution_op
      name=self.__class__.__name__,
Node: 'deep_lab_v3_3/segmentation_model_6/res_net_6/conv2d_54/Conv2D'
DNN library is not found.
	 [[{{node deep_lab_v3_3/segmentation_model_6/res_net_6/conv2d_54/Conv2D}}]] [Op:__inference_train_function_126419]
```

**Comment by qlzh727** [comment]
The runtime log I got from colab is like below, which suggest the tf version and cudnn version doesn't match. Btw, I was using tf-nightly and tf-models-nightly for testing.

```


Sep 6, 2022, 4:01:03 PM | WARNING | 2022-09-06 23:01:03.550215: W tensorflow/core/framework/op_kernel.cc:1830] OP_REQUIRES failed at conv_ops.cc:1140 : UNIMPLEMENTED: DNN library is not found.
-- | -- | --
Sep 6, 2022, 4:01:03 PM | WARNING | 2022-09-06 23:01:03.548124: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:421] Loaded runtime CuDNN library: 8.0.5 but source was compiled with: 8.1.0. CuDNN library needs to have matching major version and equal or higher minor version. If using a binary install, upgrade your CuDNN library. If building from sources, make sure the library loaded at runtime is compatible with the version specified during compile configuration.


```

**Comment by martin-gorner** [comment]
I was not using a GPU when I got this error. Just CPU.

File: setup.py [hunk 1]
```diff
@@ -34,7 +34,7 @@
     install_requires=["packaging", "absl-py", "tensorflow_datasets"],
     python_requires=">=3.7",
     extras_require={
-        "tests": ["flake8", "isort", "black", "pytest"],
+        "tests": ["flake8", "isort", "black", "pytest", "tf-models-official"],
```
Comment (bhack) [inline setup.py:37]: Is this going to change our relationship with models garden or is this just an experiment?

Comment (qlzh727) [reply setup.py:37]: This is more for an experiment for the moment.

**Comment by qlzh727** [comment]
@martin-gorner, this is quite weird. I was able to train the model e2e on a GPU instance on GCP without issue. 

The training logic is similar as yours in the colab:

```
EPOCHS = 5
model = DeepLabV3(
                classes=1,
                input_size=[*TARGET_SIZE, 3],
                backbone="resnet_50",
                backbone_weights=None,
                decoder="fpn")

    model.compile(optimizer='adam',
                  loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
                  metrics=['accuracy'])

    model_history = model.fit(training_dataset, epochs=EPOCHS,
                              steps_per_epoch=NB_IMAGES//BATCH_SIZE)      
```

And the result is like below:

```
368/368 [==============================] - 178s 381ms/step - loss: 0.4393 - accuracy: 0.8819
Epoch 2/30
368/368 [==============================] - 142s 383ms/step - loss: 0.2263 - accuracy: 0.9007
Epoch 3/30
368/368 [==============================] - 140s 382ms/step - loss: 0.1900 - accuracy: 0.9067
Epoch 4/30
368/368 [==============================] - 140s 381ms/step - loss: 0.1727 - accuracy: 0.9105
Epoch 5/30
368/368 [==============================] - 141s 382ms/step - loss: 0.1638 - accuracy: 0.9122
```

File: keras_cv/models/segmentation/deeplab/deeplab_v3.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                    f'{supported_premade_backbone}, received "{backbone}"'
+                )
+            input_spec = tf.keras.layers.InputSpec(shape=[None] + list(input_size))
+            if backbone == "resnet_50":
+                backbone = tfm.vision.backbones.ResNet(
+                    model_id=50,
+                    input_specs=input_spec,
+                    se_ratio=0.0,  # From the default Resnet config
+                    activation=activation,
+                    use_sync_bn=use_sync_bn,
```
Comment (LukeWood) [inline keras_cv/models/segmentation/deeplab/deeplab_v3.py:107]: Do we intend to integrate this with keras CV pretrained weights?

Comment (qlzh727) [reply keras_cv/models/segmentation/deeplab/deeplab_v3.py:107]: yes we should, once we have a set of backbone that is ready to use in kcv, then we should probably replace the one from MG with them.

**Comment by qlzh727** [comment]
See a demo of training result in https://colab.corp.google.com/drive/1UeTn-MJ9DW9ujy5HKVf4yK5DQ9JmK_Re#scrollTo=PEoCoZymu1nE.

**Comment by martin-gorner** [comment]
That's a very nice successful run!
I tried your notebook in my environment and unfortunately it fails. I'm not even using a GPU.

My environment is:
Tensorflow version 2.9.0
TensorFlow datasets version 4.6.0
Keras CV version 0.3.1
Numpy version 1.21.6
TF-models-official version: 2.9.0

I've noticed you have some funky google-internal imports at the beginning of your notebook. Is that needed ?

I'm using TF 2.9.0 because tf-models-official requires tensorflow-text and has a binary dependency on it so versions have to be exactly the same. The highest version available for both is 2.9.0. I'll try to bump this up and see if it helps.

**Comment by martin-gorner** [comment]
No luck trying any version above 2.9.0. Model Garden did not release a 2.10 yet and MG 2.9 force-installs TF 2.9.0 and  tensorflow-text 2.9.0. With that config I get the "Gradients do not exist " error.

Anyway, It looks like a versioning issue on my side. At least you have a working version. That's good enough to have a discussion about the UI.

**Pull request closed without merging**